lib/system/jssys.nim

```diff
@@ -476,6 +476,21 @@ proc negInt(a: int): int {.compilerproc.} =
 proc negInt64(a: int64): int64 {.compilerproc.} =
   result = a*(-1)
 
+proc nimFloatToString(a: float): cstring {.compilerproc.} =
+  ## ensures the result doesn't print like an integer, ie return 2.0, not 2
+  asm """
+    function nimOnlyDigitsOrMinus(n) {
+      return n.toString().match(/^-?\d+$/);
+    }
+    if (Number.isSafeInteger(`a`)) `result` =  `a`+".0"
```

Comment:
A combination of [toExponential](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toExponential) and [toFixed](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toFixed) might be useful here (especially for the failing `repr(3.4e38)` test, they both have more support that Number.isInteger (including IE support). It will have an effect on performance but I think if one wants that they can just `proc performantToString(n: float): cstring {.importjs: "#.toString()".}`.

`toFixed` expands the exponential up to 1e20 whereas Nim's C backend expands the exponential up to 1e15, so we have to account for that for accuracy.

```suggested
    if (`a` >= 1e15) `result` = `a`.toExponential();
    else if (Number.isInteger(`a`)) `result` = `a`.toFixed(1);
    else `result` = `a`.toString();
```

Reply:
* this would print 0.0 as 0e+0 (ok, easy to fix)
> They both have IE support.

That shouldn't be the criterion, for same reason as i gave in https://github.com/nim-lang/RFCs/issues/187#issuecomment-619561119 ; the correct way IMO is to use the more standard / newer features + provide polyfills for those so that we get the best of both worlds:
* old browsers (= IE in practice) still work thanks to our polyfill
* new browsers/node benefit from a (likely) higher performance from that builtin function

We can also use `--jsversion:es5|es6|...` to branch at CT on what's generated

* same argument holds for BigInt (but obviously that's a trickier, separate issue)

that being said, it'd be good to have c,vm,js be consistent as much as possible; where did you find those consants in nim code?

Reply:
Yeah, I don't care for IE support either but it's still good if it is supported.

I didn't look for constants I tested values, but yes that line must be the definition.

If we implement ryu in pure Nim then we won't need a JS library for it (of which there is none), or one could just use WASM.

Reply:
so looks like it's gonna be hard to match exactly what `nim c` or `nim vm` produces because it's already inconsistent across OS, see https://github.com/timotheecour/Nim/issues/138. So I'm instead leaving it at the formula from the PR which at least fixes the bug it was intended to fix (print floats so they  don't look like int's); further tweaks can be done in other PR's if needed.

(and yes, we need better wasm interop, it'd simplify lots of things, in particular performance; separate topic though; my best luck with nim+wasm was through emscripten so far)

Reply:
I didn't mean to shift the focus to consistency, correctness is best above all, which is why I'd recommend you try a hand at #7169 next.

You left a small mistake in your code, in nimOnlyDigitsOrMinus it does `n.toString()` even though the argument is a string. Beyond that, I have an alternative to my previous solution that's cleaner and focused on correctness, in case you decide against using regex.

```js
var abs;
if (`a` === 0) `result` = "0.0";
else if (!isFinite(`a`) || (abs = Math.abs(`a`), abs) >= 1e20 || abs < 1e-6 || Math.floor(`a`) !== `a`) `result` = `a`+"";
else `result` = `a`+".0";
```

There are a couple things you can do here, replace `+".0"` with toFixed(1) and `+""` with toString(), you can remove the `abs` variable and just call Math.abs twice (it's practically an OR on a bit), you can use !Number.isInteger instead of Math.floor !== a, you can get rid of the a === 0 branch and replace `abs < 1e-6` with `(a !== 0 && abs < 1e-6)`.